### PR TITLE
Issue 3908

### LINKED ISSUE #3908: Factorize the confirmExit in shell
Currently we need to reimplement the `confirmExit` on every tunnel shell
A simple solution would be to create a parent component `TunnelShell` that manage the `confirmExit` methods in the shells.
Or create a reusable function just for this method...
I'm open to ideas

### PR DESCRIPTION
**Pull request by fritzschoff**
close #3908

### CODE REVIEW

File: libs/movie/src/lib/movie/form/movie.shell.config.ts [hunk 1]
```diff
@@ -76,7 +76,7 @@ export class MovieShellConfig implements FormShellConfig<MovieControl, Movie> {
     return [onMovieChanges, onStatusChanges];
   }
 
-  async onSave({ publishing }: FormSaveOptions): Promise<any> {
+  async onSave({ publishing }: FormSaveOptions): Promise<void> {
```
Comment (pldespaigne) [inline libs/movie/src/lib/movie/form/movie.shell.config.ts:79]: There is no need to explicitly type here, InetlliSense will do it for you

File: libs/ui/src/lib/tunnel/exit/exit.component.ts [hunk 2]
```diff
@@ -1,27 +1,70 @@
...
+      this.navigate();
+    }
+    const dialogRef = this.dialog.open(TunnelConfirmComponent, {
+      width: '80%',
+      data: this.dialogText
+    });
+    this.sub = dialogRef.afterClosed().subscribe(result => {
+      if (typeof result === 'undefined') {
+        return;
+      }
```
Comment (pldespaigne) [inline libs/ui/src/lib/tunnel/exit/exit.component.ts:51]: Is this the intended behavior ?
"If the user click outside of the dialog, we neither save nor navigate"

**Review by Mussuranae — APPROVED** [review]

**Review by pldespaigne — APPROVED** [review]

File: libs/ui/src/lib/tunnel/exit/exit.component.ts (hunk 2 above)
Comment (fritzschoff) [reply libs/ui/src/lib/tunnel/exit/exit.component.ts:51]: yes we do nothing, just close modal

File: libs/ui/src/lib/tunnel/exit/exit.component.ts [hunk 3]
```diff
@@ -1,27 +1,70 @@
...
+      result ? this.save() : this.navigate()
+    });
+  }
+
+  private async save() {
+    for (const name in this.configs) {
```
Comment (GrandSchtroumpf) [inline libs/ui/src/lib/tunnel/exit/exit.component.ts:57]: Use Promise.all

File: libs/ui/src/lib/tunnel/exit/exit.component.ts [hunk 4]
```diff
@@ -1,27 +1,70 @@
...
+    @Inject(FORMS_CONFIG) private configs: ShellConfig,
+    private dialog: MatDialog
   ) { }
 
   ngOnInit() {
     this.routeBeforeTunnel = this.exitRedirect || '/c/o/';
   }
 
-  async redirect() {
+  redirect() {
```
Comment (GrandSchtroumpf) [inline libs/ui/src/lib/tunnel/exit/exit.component.ts:39]: This is not the right way to handle exit. What happens if the user click on the browser back button ?

**Review by GrandSchtroumpf — CHANGES_REQUESTED** [review]

**Pull request closed without merging**